File: benchmark/simexpal_runners/run_experiments.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <chrono>
#include <unistd.h>
#include <omp.h>
#include <nlohmann/json.hpp>

#include "Probabilistic.h"
#include "KendallTau.hpp"
#include <sstream>

#include "CommandLineParser.hpp"

using json = nlohmann::json;
using namespace std;

void print_usage() {
    cerr << "Usage: run_experiments -i <input_file> -o <output_file> -threads <int> -k <int> -delta <float> -epsilon <float> -a <algorithm> -v <implementation_language> [-s <seed>] [-d]" << endl;
    cerr << "  -d: treat graph as directed (optional)" << endl;
}

int main(int argc, char* argv[]) {
    CommandLineParser clp(argc, argv);
    
    const auto input_file = clp.value<std::string>("i", "");
    const auto output_file = clp.value<std::string>("o", "");
    
    if (input_file.empty() || output_file.empty()) {
        cerr << "Error: Input (-i) and output (-o) files are required." << endl;
        print_usage();
        return 1;
    }

    const auto threads = clp.value<int>("threads", 1);
    const auto k = clp.value<int>("k", 0);
    const auto delta = clp.value<double>("delta", 0.1);
    const auto epsilon = clp.value<double>("epsilon", 0.01);
    const auto algorithm = clp.value<std::string>("a", "kadabra");
    const auto version = clp.value<std::string>("v", "cpp");
    const auto directed = clp.value<bool>("d", false);
    if (algorithm != "kadabra") {
        cerr << "Error: C++ version only supports 'kadabra' algorithm." << endl;
        return 1;
    }
    const auto seed = clp.value<int>("s", 0);

    // Set threads
    omp_set_num_threads(threads);

    // Load graph
    auto io_start = chrono::high_resolution_clock::now();
    Probabilistic G(input_file, directed, 0); // verb = 0 (silent)
    auto io_end = chrono::high_resolution_clock::now();
    chrono::duration<double> io_diff = io_end - io_start;
    double io_time = io_diff.count();

    // Run KADABRA and measure time
    auto start = chrono::high_resolution_clock::now();
    
    if (seed != 0) {
        G.set_seed((uint32_t)seed);
    }
    
    G.run((uint32_t) k, delta, epsilon);

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> diff = end - start;
    double execution_time = diff.count();

    // Prepare JSON output
    json j;
    j["parameters"]["input_file"] = input_file;
    j["parameters"]["algorithm"] = algorithm;
    j["parameters"]["version"] = version;
    j["parameters"]["threads"] = threads;
    j["parameters"]["k"] = k;
    j["parameters"]["delta"] = delta;
    j["parameters"]["epsilon"] = epsilon;
    j["parameters"]["directed"] = directed;
    j["parameters"]["seed"] = seed;
    j["io_time_seconds"] = io_time;

    j["execution_time_seconds"] = execution_time;
    j["num_samples"] = G.get_n_pairs();

    // Helper to get graph name
    std::string graph_name;
    size_t last_slash = input_file.find_last_of('/');
    std::string filename = (last_slash == std::string::npos) ? input_file : input_file.substr(last_slash + 1);
    size_t last_dot = filename.find_last_of('.');
    if (last_dot != std::string::npos) {
        filename = filename.substr(0, last_dot);
    }
    graph_name = filename;

    std::string gt_path;
    size_t inst_pos = input_file.find("Instances/");
    if (inst_pos != std::string::npos) {
        gt_path = input_file.substr(0, inst_pos + 10) + "ground_truth/test_instances/" + graph_name + "_bet.csv";
    } else {
        gt_path = "Instances/ground_truth/test_instances/" + graph_name + "_bet.csv";
    }

    double tau_overall = NAN;
    double tau_topk = NAN;
    long long overlap = 0;
    double max_ae = NAN;
    double mae = NAN;
    double ndcg_topk = NAN;

    std::ifstream gt_file(gt_path);
    uint32_t nn = G.get_nn();
    
    if (gt_file.is_open()) {
        std::string line;
        std::getline(gt_file, line); // Skip header
        
        std::vector<std::pair<int, double>> exact_entries;
        int shift = -1;
        uint32_t max_node = 0;
        
        while (std::getline(gt_file, line)) {
            if (line.empty()) continue;
            std::stringstream ss(line);
            std::string item;
            if (std::getline(ss, item, ',')) {
                int u = std::stoi(item);
                if (std::getline(ss, item, ',')) {
                    double val = std::stod(item);
                    exact_entries.push_back({u, val});
                    if (shift == -1) {
                        shift = (u == 0) ? 1 : 0;
                    }
                    if ((uint32_t)(u + shift) > max_node) {
                        max_node = u + shift;
                    }
                }
            }
        }
        gt_file.close();

        std::vector<double> bt_exact(max_node + 1, 0.0);
        for (const auto& entry : exact_entries) {
            bt_exact[entry.first + shift] = entry.second;
        }

        std::vector<double> bt_approx(max_node + 1, 0.0);
        for (uint32_t v = 0; v < nn; ++v) {
            double cent = G.get_centrality(v);
            if (v <= max_node) {
                bt_approx[v] = cent;
            }
        }

        tau_overall = kendall_tau(bt_exact, bt_approx);

        max_ae = 0.0;
        double sum_ae = 0.0;
        for (int v = 0; v <= (int)max_node; v++) {
            double ae = std::abs(bt_exact[v] - bt_approx[v]);
            if (ae > max_ae) max_ae = ae;
            sum_ae += ae;
        }
        mae = (max_node + 1 > 0) ? sum_ae / (max_node + 1) : 0.0;

        int eval_k = (k == 0) ? 100 : k;
        if (eval_k > (int)max_node + 1) eval_k = max_node + 1;

        std::vector<int> p_exact(max_node + 1);
        std::vector<int> p_approx(max_node + 1);
        for (int i = 0; i <= (int)max_node; i++) {
            p_exact[i] = i;
            p_approx[i] = i;
        }

        std::sort(p_exact.begin(), p_exact.end(), [&](int a, int b) {
            if (bt_exact[a] != bt_exact[b]) return bt_exact[a] > bt_exact[b];
            return a < b;
        });
        std::sort(p_approx.begin(), p_approx.end(), [&](int a, int b) {
            if (bt_approx[a] != bt_approx[b]) return bt_approx[a] > bt_approx[b];
            return a < b;
        });

        std::vector<double> topk_exact_vals(eval_k);
        std::vector<double> topk_approx_vals(eval_k);
        std::vector<int> topk_exact_nodes(eval_k);
        std::vector<int> topk_approx_nodes(eval_k);
        
        for (int i = 0; i < eval_k; i++) {
            int u = p_exact[i];
            topk_exact_vals[i] = bt_exact[u];
            topk_approx_vals[i] = bt_approx[u];
            topk_exact_nodes[i] = u;
            
            int v = p_approx[i];
            topk_approx_nodes[i] = v;
        }
        
        tau_topk = kendall_tau(topk_exact_vals, topk_approx_vals);
        
        std::sort(topk_exact_nodes.begin(), topk_exact_nodes.end());
        std::sort(topk_approx_nodes.begin(), topk_approx_nodes.end());
        std::vector<int> intersection;
        std::set_intersection(topk_exact_nodes.begin(), topk_exact_nodes.end(),
                              topk_approx_nodes.begin(), topk_approx_nodes.end(),
                              std::back_inserter(intersection));
        overlap = intersection.size();

        double dcg = 0.0;
        double idcg = 0.0;
        for (int i = 0; i < eval_k; i++) {
            int u_approx = topk_approx_nodes[i];
            int u_exact = topk_exact_nodes[i];
            dcg += bt_exact[u_approx] / std::log2(i + 2.0);
            idcg += bt_exact[u_exact] / std::log2(i + 2.0);
        }
        ndcg_topk = (idcg > 0.0) ? dcg / idcg : 1.0;
    } else {
        std::cerr << "Ground truth file not found: " << gt_path << std::endl;
    }

    if (!std::isnan(tau_overall)) j["tau_overall"] = tau_overall;
    else j["tau_overall"] = nullptr;
    
    if (!std::isnan(tau_topk)) j["tau_topk"] = tau_topk;
    else j["tau_topk"] = nullptr;
    
    j["overlap_topk"] = overlap;

    if (!std::isnan(max_ae)) j["max_ae"] = max_ae;
    else j["max_ae"] = nullptr;

    if (!std::isnan(mae)) j["mae"] = mae;
    else j["mae"] = nullptr;

    if (!std::isnan(ndcg_topk)) j["ndcg_topk"] = ndcg_topk;
    else j["ndcg_topk"] = nullptr;

    // Write to output file
    ofstream o(output_file);
    if (!o.is_open()) {
        cerr << "Error: Could not open output file " << output_file << endl;
        return 1;
    }
    o << j.dump(4) << endl;
    o.close();

    cout << "Successfully wrote results to " << output_file << endl;
    return 0;
}

File: benchmark/simexpal_runners/run_experiments.jl
using ArgParse
using JSON
using Graphs
using StatsBase
import JLD2

const USE_STATIC_GRAPHS = true
if USE_STATIC_GRAPHS
    using StaticGraphs
end

# Wait, let's load what we need dynamically based on the algorithm argument or include them directly.


# In the benchmark script, they used `include` directly:
include(joinpath(@__DIR__, "..", "scripts", "BenchmarkUtils.jl"))
using .BenchmarkUtils

function parse_cmdline()
    s = ArgParseSettings()

    @add_arg_table s begin
        "--input_file", "-i"
        help = "Path to the input graph file"
        required = true
        "--output_file", "-o"
        help = "Path to the output JSON file"
        required = true
        "--threads", "-t"
        help = "Number of threads to use (Julia uses JULIA_NUM_THREADS env by default, but we accept this parameter for logging)"
        arg_type = Int
        default = 1
        "-k"
        help = "Top-K value (0 means absolute error bound for all)"
        arg_type = Int
        default = 0
        "--delta", "-d"
        help = "Confidence parameter"
        arg_type = Float64
        default = 0.1
        "--epsilon", "-e"
        help = "Error bound parameter"
        arg_type = Float64
        default = 0.01
        "-a", "--algorithm"
        help = "Algorithm to run: kadabra or brava"
        arg_type = String
        default = "kadabra"
        "-v", "--version"
        help = "Implementation version"
        arg_type = String
        default = "julia"
        "--directed"
        help = "Treat graph as directed"
        action = :store_true
        "--seed", "-s"
        help = "Random seed (0 means no seed / random)"
        arg_type = Int
        default = 0
        "--gpu"
        help = "Use GPU for BRAVA inference if available"
        action = :store_true
    end

    return parse_args(s)
end

include(joinpath(@__DIR__, "..", "..", "src", "kadabra.jl"))
include(joinpath(@__DIR__, "..", "..", "src", "BRAVAGNN.jl"))

function main()
    println("ARGS: ", ARGS)
    try
        parsed_args = parse_cmdline()

        input_file = parsed_args["input_file"]
        output_file = parsed_args["output_file"]
        threads = Threads.nthreads()
        println("Aktive Threads: ", Threads.nthreads())
        k = parsed_args["k"]
        delta = parsed_args["delta"]
        epsilon = parsed_args["epsilon"]
        algo = parsed_args["algorithm"]
        version = parsed_args["version"]
        is_directed = parsed_args["directed"]

        # ---------------------------------------------------------
        # Pre-load BRAVA model if needed
        # ---------------------------------------------------------
        brava_model = nothing
        if algo == "brava"
            weight_path = joinpath(@__DIR__, "..", "cache", "bravagnn_weights.jld2")
            if isfile(weight_path)
                Main.JLD2.@load weight_path model
                brava_model = model
            end
        end

        # ---------------------------------------------------------
        # JIT WARMUP
        # ---------------------------------------------------------
        println("Performing JIT Warmup...")
        dummy_g_raw = Main.Graphs.SimpleGraph(100, 500)
            dummy_g =
                USE_STATIC_GRAPHS ?
                (is_directed ? StaticDiGraph(dummy_g_raw) : StaticGraph(dummy_g_raw)) :
                dummy_g_raw

            if algo == "kadabra"
                Main.kadabra_centrality(
                    dummy_g,
                    k,
                    epsilon,
                    delta;
                    start_factor = 10,
                    endpoints = false,
                    rng = parsed_args["seed"] == 0 ? nothing : Main.Random.Xoshiro(parsed_args["seed"])
                )
            elseif algo == "brava"
                Main.BRAVAGNN.brava_centrality(
                    dummy_g, 
                    k, 
                    epsilon, 
                    delta;
                    model = brava_model,
                    use_gpu = parsed_args["gpu"]
                )
            elseif algo == "brandes"
                Main.Graphs.betweenness_centrality(
                    dummy_g;
                    normalize = false,
                    endpoints = false,
                )
            end
            println("Warmup done")

        # Load graph
        io_start_time = time_ns()
        g_raw = BenchmarkUtils.load_graph_from_edgelist(input_file, is_directed)
        g =
            USE_STATIC_GRAPHS ? (is_directed ? StaticDiGraph(g_raw) : StaticGraph(g_raw)) :
            g_raw
        io_end_time = time_ns()
        io_time = (io_end_time - io_start_time) / 1e9

        # JIT WARMUP: Run the algorithm on a tiny dummy graph to compile all functions
        dummy_g_raw = typeof(g_raw)(3) # Create empty graph of exact same type (e.g. SimpleGraph{Int32} vs Int64)
        Main.Graphs.add_edge!(dummy_g_raw, 1, 2)
        Main.Graphs.add_edge!(dummy_g_raw, 2, 3)
        dummy_g =
            USE_STATIC_GRAPHS ?
            (is_directed ? StaticDiGraph(dummy_g_raw) : StaticGraph(dummy_g_raw)) :
            dummy_g_raw

        if algo == "kadabra"
            
            Main.kadabra_centrality(
                dummy_g,
                k,
                epsilon,
                delta;
                start_factor = 10,
                endpoints = false,
                rng = parsed_args["seed"] == 0 ? nothing : Main.Random.Xoshiro(parsed_args["seed"])
            )
        elseif algo == "brava"
            Main.BRAVAGNN.brava_centrality(
                dummy_g, 
                k, 
                epsilon, 
                delta;
                model = brava_model
            )
        elseif algo == "brandes"
            Main.Graphs.betweenness_centrality(
                dummy_g;
                normalize = false,
                endpoints = false,
            )
        end
        # END WARMUP
        println("Warmup done")

        local lower_bounds = nothing
        local upper_bounds = nothing

        if algo == "kadabra"
            # Start timing
            start_time = time_ns()

            # Run kadabra
            res = Main.kadabra_centrality(
                g,
                k,
                epsilon,
                delta;
                start_factor = 100,
                endpoints = false,
                rng = parsed_args["seed"] == 0 ? nothing : Main.Random.Xoshiro(parsed_args["seed"])
            )
            centralities = res.centralities
            lower_bounds = res.lower_bounds
            upper_bounds = res.upper_bounds
            n_samples = res.n_samples

            end_time = time_ns()
            execution_time = (end_time - start_time) / 1e9

        elseif algo == "brava"
            # For brava
            start_time = time_ns()

            centralities, n_samples = Main.BRAVAGNN.brava_centrality(
                g, 
                k, 
                epsilon, 
                delta;
                model = brava_model,
                use_gpu = parsed_args["gpu"]
            )

            end_time = time_ns()
            execution_time = (end_time - start_time) / 1e9

        elseif algo == "brandes"
            # For brandes (exact)
            start_time = time_ns()

            centralities =
                Main.Graphs.betweenness_centrality(g; normalize = false, endpoints = false)
            n_samples = 0

            end_time = time_ns()
            execution_time = (end_time - start_time) / 1e9

        else
            error("Unknown algorithm: $algo")
        end

        # Read Ground truth
        graph_name = replace(basename(input_file), ".txt" => "")
        gt_path = joinpath(@__DIR__, "..", "..", "Instances", "ground_truth", "test_instances", "$(graph_name)_bet.csv")
        
        tau_overall = NaN
        tau_topk = NaN
        overlap = 0
        max_ae = NaN
        mae = NaN
        ndcg_topk = NaN
        
        if isfile(gt_path)
            lines = readlines(gt_path)
            if length(lines) >= 2
                shift = parse(Int, split(lines[2], ",")[1]) == 0 ? 1 : 0
                max_node = 0
                for line in lines[2:end]
                    parts = split(line, ",")
                    u = parse(Int, parts[1])
                    max_node = max(max_node, u + shift)
                end
                
                bt_exact = zeros(Float64, max_node)
                for line in lines[2:end]
                    parts = split(line, ",")
                    u = parse(Int, parts[1])
                    bt_exact[u + shift] = parse(Float64, parts[2])
                end
                
                bt_approx = zeros(Float64, max_node)
                for (v, cent) in enumerate(centralities)
                    if v <= max_node
                        bt_approx[v] = cent
                    end
                end
                
                try
                    tau_overall = corkendall(bt_exact, bt_approx)
                catch
                    tau_overall = NaN
                end
                
                max_ae = 0.0
                sum_ae = 0.0
                for v in 1:max_node
                    ae = abs(bt_exact[v] - bt_approx[v])
                    if ae > max_ae
                        max_ae = ae
                    end
                    sum_ae += ae
                end
                mae = max_node > 0 ? sum_ae / max_node : 0.0
                
                eval_k = k == 0 ? 100 : k
                eval_k = min(eval_k, max_node)
                
                p_exact = sortperm(bt_exact, rev=true)
                p_approx = sortperm(bt_approx, rev=true)
                
                topk_exact = p_exact[1:eval_k]
                topk_approx = p_approx[1:eval_k]
                
                try
                    tau_topk = corkendall(bt_exact[topk_exact], bt_approx[topk_exact])
                    overlap = length(intersect(topk_exact, topk_approx))
                catch
                    tau_topk = NaN
                end
                
                dcg = 0.0
                idcg = 0.0
                for i in 1:eval_k
                    u_approx = topk_approx[i]
                    u_exact = topk_exact[i]
                    
                    dcg += bt_exact[u_approx] / log2(i + 1)
                    idcg += bt_exact[u_exact] / log2(i + 1)
                end
                ndcg_topk = idcg > 0.0 ? dcg / idcg : 1.0
            end
        else
            println("Ground truth not found at $gt_path")
        end

        # Format JSON
        results = Dict(
            "parameters" => Dict(
                "input_file" => input_file,
                "algorithm" => algo,
                "version" => version,
                "threads" => threads,
                "k" => k,
                "delta" => delta,
                "epsilon" => epsilon,
                "directed" => is_directed,
            ),
            "execution_time_seconds" => execution_time,
            "io_time_seconds" => io_time,
            "num_samples" => n_samples,
            "tau_overall" => isnan(tau_overall) ? nothing : tau_overall,
            "tau_topk" => isnan(tau_topk) ? nothing : tau_topk,
            "overlap_topk" => overlap,
            "max_ae" => isnan(max_ae) ? nothing : max_ae,
            "mae" => isnan(mae) ? nothing : mae,
            "ndcg_topk" => isnan(ndcg_topk) ? nothing : ndcg_topk,
        )

        # Write out
        open(output_file, "w") do f
            JSON.print(f, results, 4)
        end
    catch e
        println("ERROR: ", e)
        for (exc, bt) in Base.catch_stack()
            showerror(stdout, exc, bt)
            println()
        end
        rethrow(e)
    end
end

main()
